### PR TITLE
Add ability to blocklist filepaths, ability to specify where gradio temp files are created

### PR DESCRIPTION
**Pull request by abidlabs**
Security around filepaths that Gradio exposes has been something that's come up frequently. This PR adds the ability to blocklist filepaths (and also improves the allowlist mechanism). 

Closes: #3939

Edit: I just merged in #4053, which further adds the ability to specify a location where all temporary files created by Gradio are saved using an environmental variable. I considered also making this a parameter in `launch()` but the problem is that Gradio components are the ones that save temporary files and they are not aware of the `Blocks` instance. Furthermore, the Gradio temporary file location should actually *not* be a parameter in code because it could lead to code breaking if it is uploaded from e.g. a Windows local system to a Linux system like Spaces

Closes: #3597

### CODE REVIEW

File: gradio/utils.py [hunk 1]
```diff
@@ -962,6 +962,16 @@ def abspath(path: str | Path) -> Path:
         return path.resolve()
 
 
+def is_in(path_1: str | Path, path_2: str | Path):
```
Comment (freddyaboulton) [inline gradio/utils.py:965]: I wonder if this should return true if the paths are equal - can help clean up some of the code

i.e

```
utils.is_in(abs_path, allowed_path) or abs_path == utils.abspath(allowed_path)
``

File: test/test_routes.py [hunk 2]
```diff
@@ -207,14 +228,74 @@ def test_get_file_allowed_by_file_directories(self):
...
-            file_directories=[os.path.dirname(allowed_file.name)],
+            allowed_paths=[os.path.dirname(allowed_file.name)],
         )
         client = TestClient(app)
 
         file_response = client.get(f"/file={allowed_file.name}")
         assert file_response.status_code == 200
         assert len(file_response.text) == len(media_data.BASE64_IMAGE)
 
+        app, _, _ = gr.Interface(lambda s: s.name, gr.File(), gr.File()).launch(
```
Comment (freddyaboulton) [inline test/test_routes.py:239]: nit: I would close the app after you're done with and launch the next one

**Review by freddyaboulton — APPROVED** [review]
Tested and works well! Thanks for adding the tests @abidlabs

File: CHANGELOG.md [hunk 3]
```diff
@@ -7,6 +7,7 @@
 ## New Features:
 
 - Add support for `visual-question-answering`, `document-question-answering`, and `image-to-text` using `gr.Interface.load("models/...")` and `gr.Interface.from_pipeline` by [@osanseviero](https://github.com/osanseviero) in [PR 3887](https://github.com/gradio-app/gradio/pull/3887)
+- Adds the ability to blocklist filepaths (and also improves the allowlist mechanism by [@abidlabs](https://github.com/abidlabs) in [PR 4047](https://github.com/gradio-app/gradio/pull/4047).
```
Comment (abidlabs) [inline CHANGELOG.md:10]: I should add a changelog entry for the abiliy to specify where temp files are created

**Comment by abidlabs** [comment]
Thanks for the review @freddyaboulton! Will address your suggestions and merge

**Comment by abidlabs** [comment]
Oh I think we'll need to release a new version of the gradio client library (why are the tests passing? That's weird)

**Comment by freddyaboulton** [comment]
> Oh I think we'll need to release a new version of the gradio client library (why are the tests passing? That's weird)

How come?

**Comment by abidlabs** [comment]
Oh awkward.. I have changes that I haven't pushed to this branch

The reason is that there are a bunch of other places where we create temporary files (search for `tempfile.NamedTemporaryFile` in the codebase) and I refactored all of them to use the temporary directory. For that, I created a helper function in the client library.

But for now, let me not push my changes and just remove #3592 from the list of closed issues. Will just move that to a new PR

**Pull request merged**